### PR TITLE
WL-2421: Get Contenthandler for the resource type and call copy

### PR DESCRIPTION
**Pull request by nicholaswilson100**

### CODE REVIEW

**Comment by nicholaswilson100** [comment]
Linked to https://github.com/ox-it/wl-citations/pull/106

Previous comments are here: https://github.com/ox-it/wl-kernel/pull/64
Are they still applicable?  I've reopened this request (or copied it) because I was having second thoughts about them....

In particular, I wasn't quite as sure as before that the ContentHandler code should be moved to BaseContentService.copyResource(), mainly because the current flow of control when duplicating a site does not go into that method and so what extra reason is there for moving the current flow there as well as this new code?  

The flow does go through ContentCopyImpl.copyResource() though which is where I've put this code.

File: api/src/main/java/org/sakaiproject/content/api/ContentChangeHandler.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.sakaiproject.content.api;
+
+public interface ContentChangeHandler {
+
+	/**
```
Comment (buckett) [inline api/src/main/java/org/sakaiproject/content/api/ContentChangeHandler.java:30]: Indentation.

File: kernel-impl/src/main/java/org/sakaiproject/content/impl/ResourceTypeRegistryImpl.java [hunk 2]
```diff
@@ -269,9 +275,16 @@ else if(ResourceType.MIME_TYPE_METAOBJ.equals(contentType))
 		return typeDefs;
 	}
 
-	/* (non-Javadoc)
-	 * @see org.sakaiproject.content.api.ResourceTypeRegistry#setResourceTypesForContext(java.lang.String, java.util.Map)
-	 */
+    /* (non-Javadoc)
```
Comment (buckett) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ResourceTypeRegistryImpl.java]: Indentation.

File: api/src/main/java/org/sakaiproject/content/api/ContentChangeHandler.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package org.sakaiproject.content.api;
+
+public interface ContentChangeHandler {
```
Comment (buckett) [inline api/src/main/java/org/sakaiproject/content/api/ContentChangeHandler.java:28]: New interface in the kernel should have good JavaDoc explaining what it's for.

File: api/src/main/java/org/sakaiproject/content/api/ResourceTypeRegistry.java [hunk 4]
```diff
@@ -95,14 +95,29 @@
 	 * @return
 	 */
 	public ResourceToolActionPipe newPipe(String initializationId, ResourceToolAction action);
-		
+
+    /**
+     * Register a ResourceType object to indicate that resources of that type can be defined in
+     * the Resources tool.  If the InteractionAction object is null or if the type object's getId()
```
Comment (buckett) [inline api/src/main/java/org/sakaiproject/content/api/ResourceTypeRegistry.java]: Indentation.

File: api/src/main/java/org/sakaiproject/content/api/ResourceTypeRegistry.java [hunk 5]
```diff
@@ -95,14 +95,29 @@
...
+     * method returns a null value, no type is registered.
+     * @param type
+     */
+    public void register(ResourceType type);
 	/**
 	 * Register a ResourceType object to indicate that resources of that type can be defined in  
-	 * the Resources tool.  If the InteractionAction object is null or if the type object's getId()  
+	 * the Resources tool ; and register a ContentChangeHandler with the ResourceType to deal with
+     * changing content for that type.  If the InteractionAction object is null or if the type object's getId()
```
Comment (buckett) [inline api/src/main/java/org/sakaiproject/content/api/ResourceTypeRegistry.java]: Indentation.

File: kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java [hunk 6]
```diff
@@ -158,6 +154,10 @@ private void copyResource(ContentCopyContext context, String resourceId) {
 				propsEdit.addAll(resource.getProperties());
 				hideImportedContent(resource);
 				chs.commitResource(newResource, NotificationService.NOTI_NONE);
+
+				ContentChangeHandler cch = ((DbContentService)chs).getResourceTypeRegistry().getContentChangeHandler(resource.getResourceType());
+				cch.copy(newResource);
```
Comment (buckett) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java]: cch as documented as returning `null` so there should be a `null` check here.

**Pull request merged**